### DISCUSSION TITLE
Automated Tests for Ad Server Proxy in SSAI Flow

### DISCUSSION THREAD

**Discussion by Nfrederiksen** [EyevinnOSC/community#36] [Show and tell]
Ok, so I work with a microservice that acts as a proxy between an ad server and our SSAI solution's ad-stitching service, fetching and processing VAST and VMAP files. This service retrieves ads and parses and logs metadata like creative durations, IDs, and types, making necessary tweaks based on our internal specs. Pretty cool, right? 😊

Every time I update this service, I have to ensure the end-to-end flow produces the expected VAST/VMAP output. Depending on whether a client is watching a live or VOD stream, the requests to my proxy and the ad server can vary a lot. This means I need to verify multiple GET requests with different query parameters after each update.
In the past, I manually tested each scenario, making requests to my proxy and then to the ad server. While thorough, it was super time-consuming and diluted our ad metrics with unnecessary requests.
To streamline this, I decided to make use of the Eyevinn Test Ad Server, currently available in OSC!

 🎉 Check it out here: https://app.osaas.io/dashboard/service/eyevinn-test-adserver

This tool lets me generate mock VAST and VMAP files with real media and tracking URLs, so I can test without messing with actual ad metrics. It also stores ad request data in its sessions which can be accessed with the Test Ad Server API.
There I can easily check session data to confirm my proxy service accurately constructed the correct ad server query string parameters and modified the ad servers VAST/VMAP correctly. This automation saves me time and ensures my service performs correctly without having to make wasteful ad requests to the real adserver.

![image](https://github.com/user-attachments/assets/f7817dcf-128b-48e2-b675-1b29ba43de26)

1. Service makes an ad GET request, eg.
 `https://<tas-host.com>/vmap?dur=120&uid=test1&rawQueryString=prof%3Dssai_yospace%26assetid%3D20100001%26device%3Dondomain_playse_samsungtv_app_vod%26flag%3D%2Bplay%2Bqtcb%2Bnucr%2Bemcr%2Bdtrd%2Bscpv%2Bsltp%2Bvicb%2Bslcb%26metr%3D7%26mode%3Don-demand%26resp%3Dvmap1%2Bvast4%3B` to the TAS (Test Ad Server), in the URL it has TAS-specific query parameters (to ensure a valid response) and also the query parameters it would have sent to the actual ad server.
2. TAS returns a VMAP XML file with 120 seconds of ads. The proxy is now free to go to work and parse and modify it.
3. The proxy test script would then GET the session from the API, using in this case the userID `uid` it put in the test ad request. eg.
`https://<tas-host.com>/api/v1/users/test1`
 The endpoint returns a session object that contains every query parameter sent and also the ad response. The test script cross-references the query parameters the session has with what was expected in the assertion part of the test.
4. Proxy test script analyzes the proxy output VMAP XML file and verifies that it contains all planned modifications.

**Comment by birme** [comment]
That is awesome, thank you so much for sharing this with the community!
